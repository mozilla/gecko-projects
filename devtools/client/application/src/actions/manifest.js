/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { services } = require("../modules/services");
const {
  FETCH_MANIFEST_FAILURE,
  FETCH_MANIFEST_START,
  FETCH_MANIFEST_SUCCESS,
  RESET_MANIFEST,
} = require("../constants");

function fetchManifest() {
  return async (dispatch, getState) => {
    dispatch({ type: FETCH_MANIFEST_START });
    const { manifest, errorMessage } = await services.fetchManifest();

    if (!errorMessage) {
      dispatch({ type: FETCH_MANIFEST_SUCCESS, manifest });
    } else {
      dispatch({ type: FETCH_MANIFEST_FAILURE, error: errorMessage });
    }
  };
}

function resetManifest() {
  return { type: RESET_MANIFEST };
}

module.exports = {
  fetchManifest,
  resetManifest,
};
