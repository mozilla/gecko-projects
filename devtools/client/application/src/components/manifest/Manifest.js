/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  article,
  h1,
  table,
  tbody,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);
const { l10n } = require("../../modules/l10n");

const ManifestItem = createFactory(require("./ManifestItem"));
const ManifestIssueList = createFactory(require("./ManifestIssueList"));
const ManifestSection = createFactory(require("./ManifestSection"));

/**
 * A canonical manifest, splitted in different sections
 */
class Manifest extends PureComponent {
  static get propTypes() {
    // TODO: Use well-defined types
    //       See https://bugzilla.mozilla.org/show_bug.cgi?id=1576881
    return {
      icons: PropTypes.array.isRequired,
      identity: PropTypes.array.isRequired,
      presentation: PropTypes.array.isRequired,
      validation: PropTypes.array.isRequired,
    };
  }

  renderIssueSection() {
    const { validation } = this.props;
    const shouldRender = validation && validation.length > 0;

    return shouldRender
      ? ManifestSection(
          {
            key: `manifest-section-0`,
            title: l10n.getString("manifest-item-warnings"),
          },
          ManifestIssueList({ issues: validation })
        )
      : null;
  }

  renderItemSections() {
    const { identity, icons, presentation } = this.props;

    const manifestMembers = [
      { localizationId: "manifest-item-identity", items: identity },
      { localizationId: "manifest-item-presentation", items: presentation },
      { localizationId: "manifest-item-icons", items: icons },
    ];

    return manifestMembers.map(({ localizationId, items }, index) => {
      return ManifestSection(
        {
          key: `manifest-section-${index + 1}`,
          title: l10n.getString(localizationId),
        },
        // NOTE: this table should probably be its own component, to keep
        //       the same level of abstraction as with the validation issues
        // Bug https://bugzilla.mozilla.org/show_bug.cgi?id=1577138
        table(
          {},
          tbody(
            {},
            // TODO: handle different data types for values (colors, imagesâ€¦)
            //       See https://bugzilla.mozilla.org/show_bug.cgi?id=1575529
            items.map(item => {
              const { key, value } = item;
              return ManifestItem({ label: key, key: key }, value);
            })
          )
        )
      );
    });
  }

  render() {
    return article(
      {},
      Localized(
        {
          id: "manifest-view-header",
        },
        h1({ className: "app-page__title" })
      ),
      this.renderIssueSection(),
      this.renderItemSections()
    );
  }
}

// Exports
module.exports = Manifest;
