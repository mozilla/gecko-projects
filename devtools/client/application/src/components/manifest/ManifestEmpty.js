/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  article,
  h1,
  p,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

/**
 * This component displays help information when no manifest is found for the
 * current target.
 */
class ManifestEmpty extends PureComponent {
  render() {
    return article(
      { className: "manifest-empty" },
      Localized(
        {
          id: "manifest-view-header",
        },
        h1({ className: "app-page__title" })
      ),
      Localized({ id: "manifest-non-existing" }, p({}))
    );
  }
}

// Exports
module.exports = ManifestEmpty;
