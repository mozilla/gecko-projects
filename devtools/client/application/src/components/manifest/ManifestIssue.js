/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  img,
  li,
  span,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const { MANIFEST_ISSUE_LEVELS } = require("../../constants");

/**
 * A Manifest validation issue (warning, error)
 */
class ManifestIssue extends PureComponent {
  static get propTypes() {
    return {
      className: PropTypes.string,
      level: PropTypes.oneOf(Object.values(MANIFEST_ISSUE_LEVELS)).isRequired,
      message: PropTypes.string.isRequired,
      // NOTE: we are currently ignoring the 'type' field that platform adds to
      //       errors
    };
  }

  getIconData(level) {
    switch (level) {
      case MANIFEST_ISSUE_LEVELS.WARNING:
        return {
          src: "chrome://devtools/skin/images/alert-small.svg",
          localizationId: "icon-warning",
        };
      case MANIFEST_ISSUE_LEVELS.ERROR:
      default:
        return {
          src: "chrome://devtools/skin/images/error-small.svg",
          localizationId: "icon-error",
        };
    }
  }

  render() {
    const { level, message, className } = this.props;
    const icon = this.getIconData(level);

    return li(
      { className: `manifest-warning ${className ? className : ""}` },
      Localized(
        { id: icon.localizationId, attrs: { alt: true, title: true } },
        img({
          className: `manifest-issue__icon manifest-issue__icon--${level}`,
          src: icon.src,
        })
      ),
      span({}, message)
    );
  }
}

// Exports
module.exports = ManifestIssue;
