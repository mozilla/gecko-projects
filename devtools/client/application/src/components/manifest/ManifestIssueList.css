.manifest-issues {
  list-style-type: none;
  padding-inline-start: 0;
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: var(--base-unit);
}

.manifest-issues__item {
  display: contents;
}
