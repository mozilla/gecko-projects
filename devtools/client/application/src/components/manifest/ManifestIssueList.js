/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const { ul } = require("devtools/client/shared/vendor/react-dom-factories");

const { MANIFEST_ISSUE_LEVELS } = require("../../constants");
const ManifestIssue = createFactory(require("./ManifestIssue"));

/**
 * A collection of manifest issues (errors, warnings)
 */
class ManifestIssueList extends PureComponent {
  static get propTypes() {
    // TODO: Use well-defined types
    //       See https://bugzilla.mozilla.org/show_bug.cgi?id=1576881
    return {
      issues: PropTypes.array.isRequired,
    };
  }

  // group the errors by level, and order by severity
  groupIssuesByLevel() {
    const { issues } = this.props;

    const errors = issues.filter(x => x.level === MANIFEST_ISSUE_LEVELS.ERROR);
    const warnings = issues.filter(
      x => x.level === MANIFEST_ISSUE_LEVELS.WARNING
    );

    return [errors, warnings];
  }

  render() {
    const groups = this.groupIssuesByLevel().filter(list => list.length > 0);

    return groups.map((list, listIndex) => {
      return ul(
        {
          className: "manifest-issues js-manifest-issues",
          key: `issuelist-${listIndex}`,
        },
        list.map((issue, issueIndex) =>
          ManifestIssue({
            className: "manifest-issues__item",
            key: `issue-${issueIndex}`,
            ...issue,
          })
        )
      );
    });
  }
}

// Exports
module.exports = ManifestIssueList;
