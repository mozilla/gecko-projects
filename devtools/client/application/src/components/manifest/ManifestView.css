/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.manifest-view {
  padding: calc(var(--base-unit) * 2) calc(var(--base-unit) * 4);
  border-bottom: 1px solid var(--grey-20);
  display: table;
  width: 100%;
}

.manifest-view:last-child {
  border-bottom: 0;
}

.manifest-view__title {
	font-size: var(--title-10-font-size);
  font-weight: var(--title-20-font-weight);
  text-align: left;
  padding: calc(var(--base-unit) * 2) 0;
}

.manifest-view__row {
  vertical-align: top;
}

.manifest-view__col-label {
	color: var(--grey-50);
  font-weight: var(--title-30-font-weight);
  width: calc(var(--base-unit) * 28);
  text-align: right;
}

.manifest-view__col-value {
  word-break: break-all;
}

.manifest-view__row-error .manifest-view__col-label {
  width: calc(var(--base-unit) * 6);
  text-align: left;
}
