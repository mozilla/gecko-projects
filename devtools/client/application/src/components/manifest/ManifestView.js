/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  article,
  caption,
  h1,
  table,
  tbody,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const ManifestItemIcon = createFactory(require("./ManifestItemIcon"));
const ManifestItemText = createFactory(require("./ManifestItemText"));
const ManifestItemWarning = createFactory(require("./ManifestItemWarning"));
/**
 * This component
 */
class ManifestView extends PureComponent {
  static get propTypes() {
    return {
      identity: PropTypes.object.isRequired,
      warnings: PropTypes.array.isRequired,
      icons: PropTypes.array.isRequired,
      presentation: PropTypes.object.isRequired,
    };
  }
  render() {
    const { identity, warnings, icons, presentation } = this.props;

    return article(
      {},
      Localized(
        {
          id: "manifest-view-header",
        },
        h1({ className: "app-page__title" })
      ),
      table(
        { className: "manifest-view", key: "errors-and-warnings" },
        Localized(
          { id: "manifest-item-warnings" },
          caption({ className: "manifest-view__title" })
        ),
        tbody(
          {},
          warnings.map((warning, index) =>
            ManifestItemWarning({ warning, key: `warning-${index}` })
          )
        )
      ),
      table(
        { className: "manifest-view", key: "identity" },
        Localized(
          { id: "manifest-item-identity" },
          caption({ className: "manifest-view__title" })
        ),
        tbody(
          {},
          Object.entries(identity).map(([key, value]) =>
            ManifestItemText({ name: key, val: value, key: `${key}` })
          )
        )
      ),
      table(
        { className: "manifest-view", key: "presentation" },
        Localized(
          { id: "manifest-item-presentation" },
          caption({ className: "manifest-view__title" })
        ),
        tbody(
          {},
          Object.entries(presentation).map(([key, value]) =>
            ManifestItemText({ name: key, val: value, key: `${key}` })
          )
        )
      ),
      table(
        { className: "manifest-view", key: "icons" },
        Localized(
          { id: "manifest-item-icons" },
          caption({ className: "manifest-view__title" })
        ),
        tbody(
          {},
          icons.map((icon, index) =>
            ManifestItemIcon({ icon, key: `warning-${index}` })
          )
        )
      )
    );
  }
}

// Exports
module.exports = ManifestView;
