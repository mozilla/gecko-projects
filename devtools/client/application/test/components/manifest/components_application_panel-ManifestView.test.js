/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

// Import libs
const { shallow } = require("enzyme");
const { createFactory } = require("react");

const ManifestView = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestView")
);

const { MANIFEST_DATA } = require("../../../src/constants");

// needs to move to reducer
const data = {
  warnings: MANIFEST_DATA.moz_validation,
  icons: MANIFEST_DATA.icons,
  identity: {
    name: MANIFEST_DATA.name,
    short_name: MANIFEST_DATA.short_name,
  },
  presentation: {
    display: MANIFEST_DATA.display,
    orientation: MANIFEST_DATA.orientation,
    start_url: MANIFEST_DATA.start_url,
    theme_color: MANIFEST_DATA.theme_color,
    background_color: MANIFEST_DATA.background_color,
  },
};

/**
 * Test for ManifestView.js component
 */

describe("ManifestView", () => {
  it("renders the expected snapshot", () => {
    const wrapper = shallow(
      ManifestView({
        identity: data.identity,
        warnings: data.warnings,
        icons: data.icons,
        presentation: data.presentation,
      })
    );
    expect(wrapper).toMatchSnapshot();
  });
});
