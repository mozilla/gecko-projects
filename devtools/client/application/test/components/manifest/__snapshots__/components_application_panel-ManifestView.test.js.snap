// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManifestView renders the expected snapshot 1`] = `
<article>
  <Localized
    id="manifest-view-header"
  >
    <h1
      className="app-page__title"
    />
  </Localized>
  <table
    className="manifest-view"
    key="errors-and-warnings"
  >
    <Localized
      id="manifest-item-warnings"
    >
      <caption
        className="manifest-view__title"
      />
    </Localized>
    <tbody>
      <ManifestItemWarning
        key="warning-0"
        warning={
          Object {
            "warn": "Icons item at index 0 is invalid.",
          }
        }
      />
      <ManifestItemWarning
        key="warning-1"
        warning={
          Object {
            "warn": "Icons item at index 2 is invalid. Icons item at index 2 is invalid. Icons item at index 2 is invalid. Icons item at index 2 is invalid.",
          }
        }
      />
    </tbody>
  </table>
  <table
    className="manifest-view"
    key="identity"
  >
    <Localized
      id="manifest-item-identity"
    >
      <caption
        className="manifest-view__title"
      />
    </Localized>
    <tbody>
      <ManifestItemText
        key="name"
        name="name"
        val="Name is a verrry long name and the name is longer tha you thinnk because it is loooooooooooooooooooooooooooooooooooooooooooooooong"
      />
      <ManifestItemText
        key="short_name"
        name="short_name"
        val="Na"
      />
    </tbody>
  </table>
  <table
    className="manifest-view"
    key="presentation"
  >
    <Localized
      id="manifest-item-presentation"
    >
      <caption
        className="manifest-view__title"
      />
    </Localized>
    <tbody>
      <ManifestItemText
        key="display"
        name="display"
        val="browser"
      />
      <ManifestItemText
        key="orientation"
        name="orientation"
        val="landscape"
      />
      <ManifestItemText
        key="start_url"
        name="start_url"
        val="root"
      />
      <ManifestItemText
        key="theme_color"
        name="theme_color"
        val="#345"
      />
      <ManifestItemText
        key="background_color"
        name="background_color"
        val="#F9D"
      />
    </tbody>
  </table>
  <table
    className="manifest-view"
    key="icons"
  >
    <Localized
      id="manifest-item-icons"
    >
      <caption
        className="manifest-view__title"
      />
    </Localized>
    <tbody>
      <ManifestItemIcon
        icon={
          Object {
            "size": "16x16",
            "src": "https://design.firefox.com/icons/icons/desktop/default-browser-16.svg",
            "type": "type/png",
          }
        }
        key="warning-0"
      />
      <ManifestItemIcon
        icon={
          Object {
            "size": "32x32",
            "src": "https://design.firefox.com/icons/icons/desktop/default-browser-16.svg",
            "type": "type/png",
          }
        }
        key="warning-1"
      />
      <ManifestItemIcon
        icon={
          Object {
            "size": "64x64",
            "src": "https://design.firefox.com/icons/icons/desktop/default-browser-16.svg",
            "type": "type/png",
          }
        }
        key="warning-2"
      />
    </tbody>
  </table>
</article>
`;
